v2/pb/circuit.proto

```diff
@@ -0,0 +1,60 @@
+syntax = "proto2";
+
+package circuit.pb;
+
+message HopMessage {
+  enum Type {
+    RESERVE = 0;
+    CONNECT = 1;
+    STATUS = 2;
+  }
+
+  required Type type = 1;
+
+  optional Peer peer = 2;
+  optional Reservation reservation = 3;
+  optional Limit limit = 4;
+
+  optional Status status = 5;
+}
+
+message StopMessage {
+  enum Type {
+    CONNECT = 0;
+    STATUS = 1;
+  }
+
+  required Type type = 1;
+
+  optional Peer peer = 2;
+  optional Limit limit = 3;
+
+  optional Status status = 4;
+}
+
+message Peer {
+  required bytes id = 1;
+  repeated bytes addrs = 2;
+}
+
+message Reservation {
+  optional int64 expire = 1;  // Unix expiration time (UTC)
+  repeated bytes addrs = 2;   // relay addrs for reserving peer
+  optional bytes voucher = 3; // reservation voucher
+}
+
+message Limit {
+  optional int32 duration = 1; // seconds
+  optional int64 data = 2;     // bytes
+}
+
+enum Status {
+  OK                      = 100;
+  RESERVATION_REFUSED     = 200;
```

Comment:
How about splitting `Status` into a `HopMessage` specific `Status` and a `StopMessage` specific `Status` like done with `Type`? The split would enforce protocol semantics at the datastructure level. E.g. it encodes the semantic that one could not receive a `RESERVATION_REFUSED` when sending a `StopMessage`.

Reply:
We could I guess, but there would be a lot of duplication; I don't feel strong about it nonetheless.